kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -0,0 +1,164 @@
+      final int threadPoolSize
+  ) {
+    this.streamThreadName = streamThreadName;
+    this.threadPoolSize = threadPoolSize;
+    this.threadPool = new HashMap<>(threadPoolSize);
+
+    this.logPrefix = new LogContext(String.format("stream-thread [%s]", streamThreadName));
+    this.log = logPrefix.logger(AsyncThreadPool.class);
+
+    final ProcessingQueue processingQueue = new ProcessingQueue(logPrefix);
+    this.processingQueue = processingQueue;
+    startThreads(processingQueue);
+  }
+
+  private void startThreads(final ProcessingQueue processingQueue) {
+    for (int i = 0; i < threadPoolSize; ++i) {
+      final String name = generateAsyncThreadName(streamThreadName, i);
+      final AsyncThread thread = new AsyncThread(name, processingQueue);
+
+      thread.setDaemon(true);
+      threadPool.put(name, thread);
+    }
+
+    for (final AsyncThread thread : threadPool.values()) {
+      thread.start();
+    }
+
+    log.info("Started up all {} async threads in this pool", threadPoolSize);
+  }
+
+  public void addProcessor(
+      final int partition,
+      final InternalProcessorContext<?, ?> originalContext,
+      final FinalizingQueue finalizingQueue
+  ) {
+    processingQueue.addPartition(partition);
+
+    final AsyncNodeContainer processorContainer = new AsyncNodeContainer(
+        streamThreadName,
+        partition,
+        new AsyncThreadProcessorContext<>(originalContext),
+        finalizingQueue
+    );
+
+    for (final AsyncThread thread : threadPool.values()) {
+      thread.addProcessor(processorContainer);
+    }
+  }
+
+  public void removeProcessor(
+      final int partition
+  ) {
+    processingQueue.removePartition(partition);
+
+    for (final AsyncThread thread : threadPool.values()) {
+      thread.removeProcessor(partition);
+    }
+  }
+
+  public void scheduleForProcessing(final int partition, final List<AsyncEvent> events) {
+    processingQueue.offer(partition, events);
+  }
+
+  /**
+   * @return the name for this AsyncThread, formatted by appending a unique
+   *         index i with the base name of the StreamThread with index n, ie
+   *         AsyncThread.getName() --> {clientId}-StreamThread-{n}-{i}
+   */
+  private static String generateAsyncThreadName(
+      final String streamThreadName,
+      final int asyncThreadIndex
+  ) {
+    return String.format("%s-%d", streamThreadName, asyncThreadIndex);
+  }
+
+  /**
+   * Starts up a daemon thread that will send the shutdown signal to
+   * all AsyncThreads in this pool, then wait for them to join.
+   * <p>
+   * Should be called by the StreamThread that owns this pool
+   * when it is time for the StreamThread itself to shut down.
+   * The AsyncThreadPool is tied to the lifecycle of a StreamThread's
+   * main consumer, and will be created and closed alongside it.
+   */
+  public void shutdown() {
+    final Thread shutdownThread = shutdownThread();
+    shutdownThread.setDaemon(true);
+    shutdownThread.start();
```

Comment:
is this just best effort? we don't need to wait until shut down finishes?

Reply:
Pretty much -- if it's a clean shutdown, all the processors will have flushed everything and the AsyncThreads will be sitting around with nothing to do. If it's an unclean shutdown, then it doesn't really matter anyways, right?